torch/distributed/_shard/checkpoint/storage.py

```diff
@@ -0,0 +1,89 @@
+import abc
+from typing import List, Dict
+
+from torch.futures import Future
+
+from .metadata import (
+    BytesReadRequest,
+    BytesWriteRequest,
+    Metadata,
+    TensorReadRequest,
+    TensorWriteRequest,
+)
+
+class StorageWriter(abc.ABC):
+    """
+    Interface to write to underlying storage system
+    """
+
+    @abc.abstractmethod
+    def write_bytes(self, requests: List[BytesWriteRequest]) -> Future[None]:
+        """
+        Performs a write request and returns a Future to wait on.
+        Args:
+            requests (BytesWriteRequest): see `./metadata.py`
+        """
+        pass
+
+    @abc.abstractmethod
+    def write_tensors(self, requests: List[TensorWriteRequest]) -> Future[None]:
+        """
+        Performs a write request and returns a Future to wait on.
+        Args:
+            requests (TensorWriteRequest): see `./metadata.py`
+        """
+        pass
+
+    @abc.abstractmethod
+    def write_metadata(self, metadata: Metadata) -> None:
+        """
+        Writes the metatdata.
+
+        Args:
+            metadata (Metadata): see `./metadata.py`
+        """
+        pass
+
+    def prepare_storage(self, storage_keys: Dict[str, int]) -> None:
+        """
+        This blocking call can be overwritten by the subclass.
+        It can use `storage_keys` to plan for any write preformace optimization.
+        e.g. non sequential and parallel writes.
+        By default, it does nothing
+
+        Args:
+            storage_keys (Dict[str, int]): key - handle's name. value - size
+                of the handle.
+        """
+        pass
+
+
+class StorageReader(abc.ABC):
+    """
+    Interface to read from the underlying storage system.
+    """
+
+    @abc.abstractmethod
+    def read_bytes(self, requests: List[BytesReadRequest]) -> Future[None]:
+        """
+        Read request and returns a Future to wait on.
+        Args:
+            requests (List[BytesReadRequest]): see `./metadata.py`]
+        """
+        pass
+
+    @abc.abstractmethod
+    def read_tensors(self, requests: List[TensorReadRequest]) -> Future[None]:
```

Comment:
Does this happen in practice? The recommended practice for param sharing is to reuse the param from a different module and not introduce storage aliasing.

The current behavior is undefined values because writes to the aliased storage are allowed to race between each other.

I agree that failing to load would be the most desirable outcome. We should address it on a followup PR as this doesn't happen with the models we're testing this with.

I filed this issue to track this: https://github.com/pytorch/pytorch/issues/76745